extensions/standard-processors/processors/TailFile.cpp

```diff
@@ -403,102 +435,136 @@ void TailFile::onTrigger(const std::shared_ptr<core::ProcessContext> &context, c
+                           const std::shared_ptr<core::ProcessSession> &session,
+                           const std::string &fileName,
+                           TailState &state) {
+  if (utils::file::FileUtils::file_size(state.fileNameWithPath()) < state.position_) {
+    processRotatedFiles(context, session, state);
+  }
+
+  processSingleFile(context, session, fileName, state);
+}
+
+void TailFile::processRotatedFiles(const std::shared_ptr<core::ProcessContext> &context,
+                                   const std::shared_ptr<core::ProcessSession> &session,
+                                   TailState &state) {
+    std::vector<TailState> rotated_file_states = findRotatedFiles(state);
+    for (TailState &file_state : rotated_file_states) {
+      processSingleFile(context, session, file_state.file_name_, file_state);
     }
+    state.position_ = 0;
+    state.checksum_ = 0;
+}
+
+void TailFile::processSingleFile(const std::shared_ptr<core::ProcessContext> &context,
+                                 const std::shared_ptr<core::ProcessSession> &session,
+                                 const std::string &fileName,
+                                 TailState &state) {
+  std::string full_file_name = state.fileNameWithPath();
+
+  if (utils::file::FileUtils::file_size(full_file_name) == 0u) {
+    logger_->log_warn("Unable to read file %s as it does not exist or has size zero", full_file_name);
+    return;
+  }
+  logger_->log_debug("Tailing file %s from %llu", full_file_name, state.position_);
+
+  std::size_t last_dot_position = fileName.find_last_of('.');
+  std::string baseName = fileName.substr(0, last_dot_position);
+  std::string extension = fileName.substr(last_dot_position + 1);
+
+  if (!delimiter_.empty()) {
+    char delim = delimiter_[0];
+    logger_->log_trace("Looking for delimiter 0x%X", delim);
+
+    std::vector<std::shared_ptr<FlowFileRecord>> flowFiles;
+    uint64_t checksum = state.checksum_;
+    session->import(full_file_name, flowFiles, state.position_, delim, true, &checksum);
+    logger_->log_info("%u flowfiles were received from TailFile input", flowFiles.size());
+
+    for (auto &ffr : flowFiles) {
+      logger_->log_info("TailFile %s for %u bytes", fileName, ffr->getSize());
+      std::string logName = baseName + "." + std::to_string(state.position_) + "-" +
+                            std::to_string(state.position_ + ffr->getSize() - 1) + "." + extension;
+      ffr->updateKeyedAttribute(PATH, state.path_);
+      ffr->addKeyedAttribute(ABSOLUTE_PATH, full_file_name);
+      ffr->updateKeyedAttribute(FILENAME, logName);
+      session->transfer(ffr, Success);
+      state.position_ += ffr->getSize();
+      state.timestamp_ = getTimeMillis();
+      state.checksum_ = checksum;
+      storeState(context);
+    }
+  } else {
+    std::shared_ptr<FlowFileRecord> flowFile = std::static_pointer_cast<FlowFileRecord>(session->create());
+    if (flowFile) {
```

Comment:
I think the logic is clearer this way, and the indentation level is not too bad (yet).  If more conditionals or loops are added later, we can make this change at that time.

Reply:
early return gets the worm

Reply:
To be clear, this is a question of style, and the goal is consensus. If I caused you to change anything with resentment, then that's a failure on my side. This is true for most of my suggestions, by the way.

My goal is not to avoid "too much" indentation, but to minimize the indentation level, unless a more important goal (e.g. immediately invoked lamba pattern for encapsulated side effects) warrants their use.

I consider every level of indentation a new context pushed onto the mental stack of the reader. The less context info needed to be kept in mind, the better. For this reason, I try to keep all if/for/lambda blocks as short as possible, so that for most code, less mental context needs to be kept in mind and more mental resources can be dedicated to understanding the specific business logic.

By handling the simple cases locally (like unsuccessful flow file creation at the point of error), the reader can avoid thinking about the possibility of the error, because it's "already handled" by the time he is at the business logic. When it's handled later, one needs to remember that the simple case in question (error in the example) was not yet handled.

Reply:
Don't worry, I won't implement any suggestion that I think will make the code worse.
In this case, I think both versions have merit, and with @adebreceni 's +1 (at least I think it was a +1), I have made this tiny change.

Reply:
yes I am in favor of early returns, sorry for the confusion 🙂